File: Backend/src/main/java/com/github/francisfire/anavis/controller/RequestController.java
package com.github.francisfire.anavis.controller;

import java.util.Set;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.github.francisfire.anavis.component.AccessCheckerComponent;
import com.github.francisfire.anavis.models.RequestPrenotation;
import com.github.francisfire.anavis.services.RequestServices;

@RestController
@RequestMapping("api/request")
public class RequestController {

	@Autowired
	private RequestServices requestServices;

	@SuppressWarnings("unused")
	@Autowired
	private AccessCheckerComponent accessCheckerComponent;

	@PreAuthorize("hasAuthority('DONOR')")
	@PostMapping("")
	public boolean createRequest(@RequestBody RequestPrenotation request) {
		return requestServices.addRequest(request);
	}

	@PreAuthorize("@accessCheckerComponent.sameUserId(principal, #donorMail)")
	@GetMapping("/donor/{donorMail}")
	public Set<RequestPrenotation> getRequestsByDonor(@PathVariable("donorMail") String donorMail) {
		return requestServices.getRequestsByDonor(donorMail);

	}

	@PreAuthorize("@accessCheckerComponent.sameUserId(principal, #officeMail)")
	@GetMapping("/office/{officeMail}")
	public Set<RequestPrenotation> getRequestsByOffice(@PathVariable("officeMail") String officeMail) {
		return requestServices.getRequestsByOffice(officeMail);
	}

	@PreAuthorize("@accessCheckerComponent.isRequestOwnedByDonorId(principal, #requestId) or @accessCheckerComponent.isRequestOwnedByOfficeId(principal, #requestId)")
	@GetMapping("/{requestId}")
	public RequestPrenotation getRequestById(@PathVariable("requestId") String requestId) {
		return requestServices.getRequestInstance(requestId);
	}

	@PreAuthorize("@accessCheckerComponent.isRequestOwnedByOfficeId(principal, #requestId)")
	@PutMapping("/{requestId}/approve")
	public boolean approveRequest(@PathVariable("requestId") String requestId) {
		return requestServices.approveRequest(requestId);
	}

	@PreAuthorize("@accessCheckerComponent.isRequestOwnedByOfficeId(principal, #requestId)")
	@PutMapping("/{requestId}/deny")
	public boolean denyRequest(@PathVariable("requestId") String requestId) {
		return requestServices.removeRequest(requestId);
	}

}

File: Frontend/lib/apicontrollers/request_controller.dart
import 'dart:async';
import 'dart:convert';

import 'package:anavis/models/requestprenotation.dart';
import 'package:anavis/providers/app_state.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:http/http.dart' as http;

class RequestController {
  String _ip;
  String _baseUrl;
  Map<String, String> _header;

  RequestController(BuildContext context) {
    _ip = AppState().getIp();
    _header = AppState().getHttpHeaders();

    _baseUrl = "http://$_ip:8080/api/request";
  }

  Future<String> createRequest(RequestPrenotation request) async {
    http.Response res = await http.post(
      Uri.encodeFull("$_baseUrl"),
      body: json.encode({
        "id": request.getId(),
        "officeMail": request.getOfficeMail(),
        "donorMail": request.getDonorMail(),
        "hour": request.getHour()
      }),
      headers: this._header,
    );
    return res.body;
  }

  Future<String> getRequestsByDonor(String donorMail) async {
    http.Response res = await http.get(
      "$_baseUrl/donor/$donorMail",
      headers: this._header,
    );
    return res.body;
  }

  Future<String> getRequestsByOffice(String officeMail) async {
    http.Response res = await http.get(
      "$_baseUrl/office/$officeMail",
      headers: this._header,
    );
    return res.body;
  }

  Future<String> getRequestById(String requestId) async {
    http.Response res = await http.get(
      "$_baseUrl/$requestId",
      headers: this._header,
    );
    return res.body;
  }

  Future<String> approveRequest(String requestId) async {
    http.Response res = await http.put(
      "$_baseUrl/$requestId/approve",
      headers: this._header,
    );
    return res.body;
  }

  Future<String> denyRequest(String requestId) async {
    http.Response res = await http.put(
      "$_baseUrl/$requestId/deny",
      headers: this._header,
    );
    return res.body;
  }
}
